content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md

```diff
@@ -0,0 +1,115 @@
+secrets, or volumes. This requires a large bandwidth to access kube-apiserver.
+
+In versions prior to v1.27, the default `kubeAPIQPS` was 5 and `kubeAPIBurst` was 10. However,
+Kubelet has increased these defaults to 50 and 100 respectively for better performance during
+pod startup. It's worth noting that this isn't the only reason why we've bumped up the API QPS
+limits for Kubelet.
+
+1. It has a potential to be hugely throttled now (default QPS = 5)
+2. In large clusters they can generate significant load anyway as there are a lot of them
+3. They have a dedicated PriorityLevel and FlowSchema that we can easily control
+
+In old days, we always encounter `volume mount timeout` on kubelet in node with more than 50 pods
+during pod start up. We suggest our customers to bump `kubeAPIQPS` to 20 and `kubeAPIBurst` to 40
+and especially with customers that is using bare-metal(physical machines).
+
+More detials can be found in the KEP <https://kep.k8s.io/1040> and the pull request [#116121](https://github.com/kubernetes/kubernetes/pull/116121).
+
+## Event triggered updates to container status
+
+`Evented PLEG` (PLEG is short for "Pod Lifecycle Event Generator") is set to be in beta for v1.27,
+disabled by default (needs to be enabled explicitly). It aids pod startup by reducing unnecessary
+work during inactivity through replacing periodic polling. The current `Generic PLEG` incurs
+non-negligible overhead due to frequent container status polling, which worsens with Kubelet's
+parallelism, leading to poor performance and reliability issues. For these reasons, it is
+recommended that we use Evented PLEG instead. 
+
+Further details can be found in the KEP <https://kep.k8s.io/3386>.
+
+## Raise your pod resource limit if needed
+
+During start-up, some pods may consume a considerable amount of CPU or memory. If the CPU limit is
+low, this can significantly slow down the pod start-up process. To improve the memory management,
+Kubernetes v1.22 introduced a feature gate called MemoryQoS to kubelet. This feature enables
+kubelet to set memory QoS at container, pod, and QoS levels for better protection and guaranteed
+quality of memory when running with cgroups v2. Although it has benefits, it is possible that
+enabling this feature gate may affect the start-up speed of the pod if the pod startup consumes
+a large amount of memory.
+
+Kubelet configuration now includes `memoryThrottlingFactor`. This factor is multiplied by
+the memory limit or node allocatable memory to set the cgroupv2 memory.high value for enforcing
+MemoryQoS. Decreasing this factor sets a lower high limit for container cgroups, increasing reclaim
+pressure. Increasing this factor will put less reclaim pressure. The default value is 0.8 initially
+and will change to 0.9 in Kubernetes v1.27. This parameter adjustment can reduce the potential
+impact of this feature on pod startup speed.
+
+Further details can be found in the KEP <https://kep.k8s.io/2570>.
```

Comment:
I think https://github.com/kubernetes/website/pull/39853 covers this. I don't mind if we merge the two articles; that's even a good thing if we do. However, we need to manage how that happens and check if the other authors are OK with the idea.

Reply:
/cc @ndixita